src/xdp-app-info.c

```diff
@@ -923,6 +923,64 @@ xdp_invocation_ensure_app_info_sync (GDBusMethodInvocation  *invocation,
                                               error);
 }
 
+static XdpAppInfo *
+transition_host_app_info_to_registered  (GDBusMethodInvocation  *invocation,
+                                         const char             *app_id,
+                                         GCancellable           *cancellable,
+                                         GError                **error)
+{
+  GDBusConnection *connection = g_dbus_method_invocation_get_connection (invocation);
+  const char *sender = g_dbus_method_invocation_get_sender (invocation);
+  g_autoptr(XdpAppInfo) registered_app_info = NULL;
+  XdpAppInfoHost *app_info_host;
+  XdpAppInfo *app_info;
+
+  g_assert (app_info_by_unique_name != NULL);
+
+  G_LOCK (app_infos);
+  app_info = g_hash_table_lookup (app_info_by_unique_name, sender);
+  G_UNLOCK (app_infos);
+
+  g_assert (XDP_IS_APP_INFO (app_info));
+
+  if (!XDP_IS_APP_INFO_HOST (app_info))
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                   "App is not a host system app");
+      return NULL;
+    }
+
+  app_info_host = XDP_APP_INFO_HOST (app_info);
+
+  if (xdp_app_info_host_is_registered (app_info_host))
+    {
+      g_set_error (error, G_IO_ERROR, G_IO_ERROR_FAILED,
+                   "Connection already associated and registered with an application ID");
+      return NULL;
+    }
+
+  registered_app_info = maybe_create_registered_test_app_info (app_id);
+
+  if (!registered_app_info)
+    {
+      g_autofd int pidfd = -1;
+      uint32_t pid;
+
+      if (!xdp_connection_get_pidfd (connection, sender, cancellable, &pidfd, &pid, error))
+        return NULL;
+
+      registered_app_info = xdp_app_info_host_new_registered (pidfd, app_id, error);
+      if (!registered_app_info)
+        return NULL;
+    }
+
+  g_debug ("Replacing host app info with a registered one for '%s'", xdp_app_info_get_id (app_info));
+
+  cache_insert_app_info (sender, app_info);
```

Comment:
Two things:

It should be `registered_app_info` and not `app_info` and also a new function like `cache_replace_app_info()` should be added, because from my understanding `g_hash_table_insert()` doesn't replace existing item so one has to use `g_hash_table_replace()`.

Reply:
`g_hash_table_insert` replaces the value, and unrefs the previous value